datafusion/physical-expr/src/utils.rs

```diff
@@ -235,6 +268,186 @@ pub fn ordering_satisfy_concrete<F: FnOnce() -> EquivalenceProperties>(
+                    })
+                    .collect::<Vec<_>>();
+                let normalized_provided = provided
+                    .iter()
+                    .map(|e| {
+                        normalize_sort_requirement_with_equivalence_properties(
+                            e.clone(),
+                            eq_classes,
+                        )
+                    })
+                    .collect::<Vec<_>>();
+                normalized_required
+                    .iter()
+                    .zip(normalized_provided.iter())
+                    .all(|(req, pro)| pro.compatible(req))
+            } else {
+                false
+            }
+        }
+    }
+}
+
+pub fn map_columns_before_projection(
+    parent_required: &[Arc<dyn PhysicalExpr>],
+    proj_exprs: &[(Arc<dyn PhysicalExpr>, String)],
+) -> Vec<Arc<dyn PhysicalExpr>> {
+    let mut column_mapping = HashMap::new();
+    for (expression, name) in proj_exprs.iter() {
+        if let Some(column) = expression.as_any().downcast_ref::<Column>() {
+            column_mapping.insert(name.clone(), column.clone());
+        };
+    }
+    let new_required: Vec<Arc<dyn PhysicalExpr>> = parent_required
+        .iter()
+        .filter_map(|r| {
+            if let Some(column) = r.as_any().downcast_ref::<Column>() {
+                column_mapping.get(column.name())
+            } else {
+                None
+            }
+        })
+        .map(|e| Arc::new(e.clone()) as Arc<dyn PhysicalExpr>)
+        .collect::<Vec<_>>();
+    new_required
+}
+
+pub fn map_requirement_before_projection(
+    parent_required: Option<&[PhysicalSortRequirements]>,
+    proj_exprs: &[(Arc<dyn PhysicalExpr>, String)],
+) -> Option<Vec<PhysicalSortRequirements>> {
+    if let Some(requirement) = parent_required {
+        let required_expr = create_sort_expr_from_requirement(requirement)
+            .iter()
+            .map(|sort_expr| sort_expr.expr.clone())
+            .collect::<Vec<_>>();
+        let new_exprs = map_columns_before_projection(&required_expr, proj_exprs);
+        if new_exprs.len() == requirement.len() {
+            let new_request = new_exprs
+                .iter()
+                .zip(requirement.iter())
+                .map(|(new, old)| PhysicalSortRequirements {
+                    expr: new.clone(),
+                    sort_options: old.sort_options,
+                })
+                .collect::<Vec<_>>();
+            Some(new_request)
+        } else {
+            None
+        }
+    } else {
+        None
+    }
+}
+
+pub fn create_sort_expr_from_requirement(
+    required: &[PhysicalSortRequirements],
+) -> Vec<PhysicalSortExpr> {
+    let parent_required_expr = required
+        .iter()
+        .map(|prop| {
+            if prop.sort_options.is_some() {
+                PhysicalSortExpr {
+                    expr: prop.expr.clone(),
+                    options: prop.sort_options.unwrap(),
+                }
+            } else {
```

Comment:
I think we can use `If let Some(sort_options) = prop.sort_options`  idiom here. This would remove `.unwrap()`